### ISSUE TITLE
Submitting forms

### ISSUE THREAD

**Issue by dy** [dai-shi/react-hooks-async#6]
Hi @dai-shi!
Just found myself puzzled with how to organize sending forms along with [react-use-form-state](https://ghub.io/react-use-form-state).
That would simplify work if there was some example of that use-case.

**Comment by dy** [comment]
Oh! There is undocumented `start` property returned from `useAsyncTask`! That's what I need.

**Comment by dai-shi** [comment]
Yeah, you can use `start`. The example is [here](https://github.com/dai-shi/react-hooks-async/tree/master/examples/03_startbutton).

The reason why it's not documented is it's not the main use case.

If you are only calling an async function from an event handler, you can just define a function. There's nothing special this library can help except for combining multiple async functions.

So, I'm interested in your use case. Please let me know how you used the library. Thanks.

**Comment by dy** [comment]
I basically submit form in `onSubmit` with [react-use-form-state](https://ghub.io/react-use-form-state) hook:

```jsx
import React from 'react'
import { connect } from 'unistore/react'
import { useFormState } from 'react-use-form-state'
import { useAsyncTask, useAsyncRun } from 'react-hooks-async'

import { actions } from '../store'

export default connect(null, actions)(function SignIn (props) {
  const [formState, {text, email, password}] = useFormState()

  // NOTE: in some reason, if not to pass `[formState.values]`, useAsyncTask causes infinite rerendering
  const signInTask = useAsyncTask(() => props.signIn(formState.values), [formState.values])

  return (
        <form
          className="sign-in-form"
          autoComplete='on'
          method='POST'
          action="/login"
          onSubmit={(e) => {
            e.preventDefault()

            // NOTE: If I get it, there should be useAsyncRun, but react raises error
            // So, documenting that should be apparent, that is quite standard use-case
            signInTask.start()
          }}
        >
          <TextField
            id="merchantAccountNumber"
            autoComplete="username"
            label={"Merchant Id"}
            required
            margin="normal"
            variant="outlined"
            helperText={formState.errors.merchantAccountNumber}
            {...text('merchantAccountNumber')}
            disabled={signInTask.started && signInTask.pending}
          />
          <TextField
            id="email"
            autoComplete="email"
            label={"Email"}
            required
            margin="normal"
            variant="outlined"
            helperText={formState.errors.email}
            {...email('email')}
            disabled={signInTask.started && signInTask.pending}
          />
          <TextField
            id="password"
            autoComplete="current-password"
            label={"Password"}
            required
            helperText={formState.errors.password}
            margin="normal"
            variant="outlined"
            {...password('password')}
            disabled={signInTask.started && signInTask.pending}
          />
          <Button
            variant="contained"
            color="primary"
            size="large"
            type="submit"
            disabled={!!signInTask.error || (signInTask.started && signInTask.pending)}
          >
            Sign In
          </Button>
        </form>
  )
})
```

If you could comment the solution, I'd greatly appreciate @dai-shi

**Comment by dai-shi** [comment]
Well, I wouldn't use `react-hooks-async` for this purpose. It's much simpler without it.

```javascript
  const [formState, {text, email, password}] = useFormState();
  const [signInStatus, setSignInStatus] = useState('ready');
  const isMounted = useRef(true);
  useEffect(() => () => { isMounted.current = false; }, []); // it's a bit tricky
  const startSignIn = async () => {
    setSignInStatus('pending');
    try {
      await props.signIn(formState.values);
      if (isMounted.current) setSignInStatus('done');
    } catch (e) {
      if (isMounted.current) setSignInStatus('error');
    }
  };
```
You can make this as a custom hook.

The whole point of `react-hooks-async` is abortability. So, useAsyncTask should be used with AbortController.

**Comment by dy** [comment]
But isn't this almost the same as `useAsyncTask`? I believe `useAsyncTask` provides nicer abstraction over handling flags/errors/status, less congnitive load with implementation details in code - hardly can imagine managing ~3-4 sets of flags for 4 requests.

**Comment by dy** [comment]
For me useAsyncTask seems nicer abstraction, hiding flags/implementation details. That's less error-prone. If we have 10 forms with junior devs, that can be implementation mess.

**Comment by dai-shi** [comment]
> nicer abstraction

Yeah, I actually agree with that.

**Comment by dai-shi** [comment]
So, what you pointed out is:
- a) this library is usable even without AbortController
- b) this library is usable without useEffect (which is in the example, though)

I wonder how I should describe this.

There's a minor issue which might be critical. A `task` created by useAsyncTask is empty at first.
By empty, I mean `task.start` is `null`. It's available only in the second render. So, technically without checking the existence of `task.start`, it may cause an error.

What you can do is:
```javascript
const signInTask = useAsyncTask(...);
if (!signInTask.start) return null; // avoid showing the form in the first render
```
and it's safe.

**Comment by dy** [comment]
Oh, I see. That’s a bit unexpected.
Is there a reason that can’t be fixed? This and infinite loop are quite
annoying details.

**Comment by dai-shi** [comment]
> Is there a reason that can’t be fixed?

Past implementation dealt with it. However, it was a very complex implementation with useRef, and I felt like it may not work correctly in concurrent mode in some edge cases. The current implementation is much cleaner in that sense. This library is meant to use with useEffect (= useAsyncRun), and it's not a big problem with that use case.

> infinite loop

Deps is required somewhere anyway for the useEffect use case.

> nicer abstraction

So, that means there should be a better abstraction for your use case.

```javascript
const initialState = {
  started: false,
  pending: true,
  error: null,
  result: null,
};

const reducer = (state, action) => {
  ...
};

const useAsyncTask = (func) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const start = async () => {
    ...
  };
  return { ...state, start };
};
```

**Comment by dy** [comment]
Yes, AsyncTask abstraction is self-sufficient, just checked [Android AsyncTask](https://developer.android.com/reference/android/os/AsyncTask)/[Q](https://github.com/kriskowal/q/wiki/API-Reference) API, the flow is quite similar.

Something like:

```js
let {result, error, start, abort, started, pending, resolved, rejected, aborted} = useAsyncTask(promise, deps)
```

**Comment by dai-shi** [comment]
As I got several requests since then, I changed my mind and redesigned it.
Now, it will support the non-useEffect use case.
If you are still interested, please try #18.

Discussion should be merged to #17. Closing this issue.
